tests/wasi_serialization.rs

```diff
@@ -1,65 +1,70 @@
+    #[test]
+    fn serializing_works() {
+        let args = vec![
+            b"program_name".into_iter().cloned().collect(),
+            b"arg1".into_iter().cloned().collect(),
+        ];
+        let envs = vec![
+            b"PATH=/bin".into_iter().cloned().collect(),
+            b"GOROOT=$HOME/.cargo/bin".into_iter().cloned().collect(),
+        ];
+        let wasm_binary = include_bytes!("wasi_test_resources/unstable/fd_read.wasm");
+        let module = compile_with(&wasm_binary[..], &*get_compiler())
+            .map_err(|e| format!("Can't compile module: {:?}", e))
+            .unwrap();
 
-    let state_bytes = {
-        let instance = module.instantiate(&import_object).unwrap();
+        let wasi_version = get_wasi_version(&module, true).expect("WASI module");
+        let import_object = generate_import_object_for_version(
+            wasi_version,
+            args.clone(),
+            envs.clone(),
+            vec![],
+            vec![(
+                ".".to_string(),
+                std::path::PathBuf::from("wasi_test_resources/test_fs/hamlet"),
+            )],
+        );
 
-        let start: Func<(), ()> = instance.exports.get("_start").unwrap();
-        start.call().unwrap();
-        let state = get_wasi_state(instance.context());
+        let state_bytes = {
+            let instance = module.instantiate(&import_object).unwrap();
 
-        assert_eq!(state.args, args);
-        assert_eq!(state.envs, envs);
-        let bytes = state.freeze().unwrap();
+            let start: Func<(), ()> = instance.exports.get("_start").unwrap();
+            start.call().unwrap();
+            let state = get_wasi_state(instance.context());
 
-        bytes
-    };
+            assert_eq!(state.args, args);
+            assert_eq!(state.envs, envs);
+            let bytes = state.freeze().unwrap();
 
-    let mut instance = module.instantiate(&import_object).unwrap();
+            bytes
+        };
 
-    let wasi_state = Box::new(WasiState::unfreeze(&state_bytes).unwrap());
+        let mut instance = module.instantiate(&import_object).unwrap();
 
-    instance.context_mut().data = Box::into_raw(wasi_state) as *mut c_void;
+        let wasi_state = Box::new(WasiState::unfreeze(&state_bytes).unwrap());
 
-    let second_entry: Func<(), i32> = instance.exports.get("second_entry").unwrap();
-    let result = second_entry.call().unwrap();
-    assert_eq!(result, true as i32);
-}
+        instance.context_mut().data = Box::into_raw(wasi_state) as *mut c_void;
+
+        let second_entry: Func<(), i32> = instance.exports.get("second_entry").unwrap();
+        let result = second_entry.call().unwrap();
+        assert_eq!(result, true as i32);
+    }
 
-#[allow(clippy::mut_from_ref)]
-pub(crate) fn get_wasi_state(ctx: &Ctx) -> &mut WasiState {
-    unsafe { state::get_wasi_state(&mut *(ctx as *const Ctx as *mut Ctx)) }
+    #[allow(clippy::mut_from_ref)]
+    pub(crate) fn get_wasi_state(ctx: &Ctx) -> &mut WasiState {
```

Comment:
Note to self, I think this can be removed now and it definitely should be!

Or actually, it looks like we may be able to change `ctx: &Ctx` to `ctx: &mut Ctx` and remove the lint rule above

Reply:
Fixed!